### PR TITLE
[expo-print][ios] Convert expo-print to swift

### LINKED ISSUE #19399: iOS 16 Multiple Print Bug
### Summary

I have two React Native applications that use `expo-print` for label printing (i.e. with Brother Label Printer QL-810W). I noticed recently that when I print multiple times in either application, the label printing is inconsistent. It will print perfectly correct once, and the next time it prints off-center and multiple blank labels (I only ever want one label printed at a time). The way I print is to render what I want to print in React Native, then take a screenshot of that component, and convert it to a base64 image, which is then sent to AirPrint.

I am almost certain this is an iOS 16 issue, because I haven't made any changes to one of these apps in a while, and printing was stable until recently. The issue only happens on iOS 16 devices. I tested on Android and iOS 15 and both are still stable. I also tested with multiple physical iOS 16 devices and an iOS 16 simulator. My team has tested this issue against several different printers, and we are seeing it everywhere. I am not sure whether this is an expo bug or an Apple bug, but I appreciate any help.

Here is a picture of what I am experiencing. Again -- these are the exact same images, printed one after the other to my label printer:
IMG_4337

### What platform(s) does this occur on?

iOS

### Environment

```
expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.5.1
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 16.13.1 - /usr/local/bin/node
      Yarn: 1.22.17 - /usr/local/bin/yarn
      npm: 8.5.2 - /usr/local/bin/npm
      Watchman: 2022.03.14.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.11.2 - /Users/username/.rvm/gems/ruby-3.0.0/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    IDEs:
      Android Studio: 2021.2 AI-212.5712.43.2112.8512546
      Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
    npmPackages:
      expo: 46.0.7 => 46.0.7 
      react: 18.0.0 => 18.0.0 
      react-dom: 18.0.0 => 18.0.0 
      react-native: 0.69.4 => 0.69.4 
    npmGlobalPackages:
      eas-cli: 0.47.0
      expo-cli: 5.0.3
    Expo Workflow: bare
```

### Minimal reproducible example

```
const imgHtml = 
`<html>
    <head>
	    <meta name="viewport" content="width=device-width,initial-scale=1,user-scalable=yes" />
	    <style>
		    @page {
			    margin: 0;
		    }
		    body {
			    height: 1.4in;
			    width: 3.5in;
			    margin: 0;
			    padding: 0;
			    overflow: hidden;
			    align-items: center;
			    justify-content: center;
		    }
		    img {
			    height: 100%;
			    width: 100%;
		    }
	    </style>
    </head>
    <body>
	    <img src="data:image/png;base64,${uri}" />
    </body>
</html>`;

await Print.selectPrinterAsync().then(async (printer) => {
    await Print.printAsync({
	    html: imgHtml,
	    margins: {
		    left: 0,
		    right: 0,
		    top: 0,
		    bottom: 0
	    },
	    useMarkupFormatter: true,
	    orientation: "landscape",
	    printerUrl: printer.url
    }).then(() => {
	    resolve(true);
    }).catch((err_1) => {
	    resolve(false);
    });
}).catch((err_2) => {
    resolve(false);
});
```

### PR DESCRIPTION
**Pull request by behenate**
# Why

Ported expo-print to swift on iOS, fixed https://github.com/expo/expo/issues/19399

# How

Ported all of the existing files to new module API, cached previously used UIPrinters as a workaround for the bug (it's an iOS bug).

# Test Plan

Tested all features on a physical printer and print to file from iOS 16.2 simulator.

### CODE REVIEW

**Comment by behenate** [comment]
I removed repeating warnings from expo bot, (redundant optional initialization  seems to be necessary in this case)

File: packages/expo-print/expo-module.config.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "platforms": ["ios", "android"],
+    "ios": {
+      "modules": ["ExpoPrintModule"]
+    }
+}
```
Comment (lukmccall) [inline packages/expo-print/expo-module.config.json:6]:
```suggestion
}

```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import ExpoModulesCore
+
+internal class PrinterPickerException: GenericException<String?> {
+  override var reason: String {
+    "There was a problem with the printer picker: '\(param ?? "nil")'"
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:5]:
```suggestion
    "There was a problem with the printer picker: '\(param ?? "unknown error")'"
```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+internal class ViewControllerNotFoundException: Exception {
+  override var reason: String {
+    "Could not find the current view controller."
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:11]:
```suggestion
    "Could not find the current view controller"
```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+internal class PickerCancelledException: Exception {
+  override var reason: String {
+    "Printer picker has been cancelled."
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:17]:
```suggestion
    "Printer picker has been canceled"
```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+internal class ViewControllerNotFoundException: Exception {
+  override var reason: String {
+    "Could not find the current view controller."
+  }
+}
+
+internal class PickerCancelledException: Exception {
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:15]:
```suggestion
internal class PickerCanceledException: Exception {
```
We prefere the American version of that word.

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  override var reason: String {
+    "Printer picker has been cancelled."
+  }
+}
+
+internal class PdfNotRenderedException: GenericException<String?> {
+  override var reason: String {
+    "Error occurred while printing to PDF: '\(param ?? "nil")'"
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:23]:
```suggestion
    "Error occurred while printing to PDF: '\(param ?? "unknown error")'"
```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+internal class PdfSavingException: Exception {
+  override var reason: String {
+    "Error occurred while saving the PDF."
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:29]:
```suggestion
    "Error occurred while saving the PDF"
```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+internal class UnsupportedFormatException: GenericException<String?> {
+  override var reason: String {
+    "Given format '\(param ?? "nil")' is not supported."
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:35]:
```suggestion
    "Given format '\(param ?? "unknown")' is not supported"
```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+internal class InvalidUrlException: Exception {
+  override var reason: String {
+    "The specified url is not valid for printing"
+  }
+}
+
+internal class NoPrintDataException: Exception {
+  override var reason: String {
+    "No data to print. You must specify `uri` or `html` option."
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:53]:
```suggestion
    "No data to print. You must specify `uri` or `html` option"
```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+internal class PrintingJobFailedException: GenericException<String?> {
+  override var reason: String {
+    "Printing job encountered an error: '\(param ?? "nil")'"
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:59]:
```suggestion
    "Printing job encountered an error: '\(param ?? "unknown error")'"
```

File: packages/expo-print/ios/ExpoPrintExceptions.swift [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+internal class PrintIncompleteException: Exception {
+  override var reason: String {
+    "Printing did not complete."
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintExceptions.swift:65]:
```suggestion
    "Printing did not complete"
```

File: packages/expo-print/ios/ExpoPrintToFile.swift [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+public class ExpoPrintToFile {
+  var renderTasks: [ExpoWKPDFRenderer] = []
+  let appContext: AppContext?
+
+  init(appContext: AppContext?) {
+    self.appContext = appContext
+  }
+
+  func printToFile(promise: Promise, options: PrintOptions) {
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintToFile.swift:11]:
```suggestion
  func printToFile(options: PrintOptions, promise: Promise) {
```
That order is more natural. The promise should always be last.

File: packages/expo-print/ios/ExpoPrintToFile.swift [hunk 13]
```diff
@@ -0,0 +1,103 @@
+import ExpoModulesCore
+
+public class ExpoPrintToFile {
+  var renderTasks: [ExpoWKPDFRenderer] = []
+  let appContext: AppContext?
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintToFile.swift:5]:
```suggestion
  weak let appContext: AppContext?
```
The current version shouldn't leak, but to be safe, it's better to make appContext a weak reference.

File: packages/expo-print/ios/ExpoPrintToFile.swift [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        let base64Data = options.base64 ? data.base64EncodedString(options: Data.Base64EncodingOptions.endLineWithLineFeed) : nil
+        let result = FilePrintResult(uri: uri.absoluteString, numberOfPages: pagesNumber, base64: base64Data)
+        promise.resolve(result)
+      } catch let error {
+        promise.reject(PdfSavingException())
+      }
+    }
+
+    if options.useMarkupFormatter {
+      ExpoPrintToFile.pdfWithHtmlMarkupFormatter(htmlString: htmlString, pageSize: pageSize, printableRect: printableRect, onFinished: completionHandler)
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintToFile.swift:53]:
```suggestion
      ExpoPrintToFile.pdfWithHtmlMarkupFormatter(
        htmlString: htmlString, 
        pageSize: pageSize, 
        printableRect: printableRect,
        onFinished: completionHandler
      )
```

File: packages/expo-print/ios/ExpoPrintToFile.swift [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    } else {
+      let renderTask = ExpoWKPDFRenderer(htmlString: htmlString, pageSize: pageSize, printableRect: printableRect, completionHandler: completionHandler)
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintToFile.swift:55]:
```suggestion
      let renderTask = ExpoWKPDFRenderer(
        htmlString: htmlString, 
        pageSize: pageSize, 
        printableRect: printableRect, 
        completionHandler: completionHandler
      )
```

File: packages/expo-print/ios/ExpoPrintWithPrinter.swift [hunk 16]
```diff
@@ -0,0 +1,149 @@
+import ExpoModulesCore
+
+public class ExpoPrintWithPrinter {
+  var renderTasks: [ExpoWKPDFRenderer] = []
+  var cachedPrinters: [String: UIPrinter] = [:]
+  let delegate: ExpoPrintModuleDelegate
+  let appContext: AppContext?
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintWithPrinter.swift:7]:
```suggestion
```
seems to be unused 🤔

File: packages/expo-print/ios/ExpoPrintWithPrinter.swift [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+
+  init(delegate: ExpoPrintModuleDelegate, appContext: AppContext?) {
+    self.delegate = delegate
+    self.appContext = appContext
+  }
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintWithPrinter.swift:12]:
```suggestion
  init(delegate: ExpoPrintModuleDelegate) {
    self.delegate = delegate
  }
```

File: packages/expo-print/ios/ExpoPrintWithPrinter.swift [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+
+  func startPrint(promise: Promise, options: PrintOptions) {
```
Comment (lukmccall) [inline packages/expo-print/ios/ExpoPrintWithPrinter.swift:14]:
```suggestion
  func startPrint(options: PrintOptions, promise: Promise) {
```

File: packages/expo-print/ios/PrintResponse.swift [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+
+internal struct FilePrintResult: Record {
+  @Field
+  var uri: String = ""
+
+  @Field
+  var numberOfPages: Int = 0
+
+  @Field
+  var base64: String? = nil
```
Comment (lukmccall) [inline packages/expo-print/ios/PrintResponse.swift:11]: Disable the swiftlint for that line and add a comment why do you that

**Review by lukmccall — APPROVED** [review]
🔥🔥🔥

**Pull request merged**